Update symbolication.md (#2630)
saucelabs/sauce-docs: docs/error-reporting/project-setup/symbolication.md

import TabItem from '@theme/TabItem';
import useBaseUrl from '@docusaurus/useBaseUrl';

> ## Introdction

Reviewer: typo

> Debug symbols are needed when crashes are reported from binaries that have information such as function names and line numbers stripped away. Debug symbols allow Backtrace to provide human-readable callstacks and deduplicate effectively. Symbols can be uploaded directly via the UI or API, or retrieved from a managed private symbol server. Symbols are compilation specific and are applied to crash reports based on matching debug or code ID’s and filename. 

Reviewer: ```suggestion
Debug symbols are needed when crashes are reported from binaries that have information such as function names and line numbers stripped away. Debug symbols allow Backtrace to provide human-readable callstacks and effectively duplicate error reports. Symbols can be uploaded via the UI or API, or retrieved from a managed private symbol server. Symbols are compilation specific and are applied to crash reports based on matching debug or code ID’s and filename. 
```

<b>ELF file with debug info </b>

To generate an ELF file with debug info, refer to compiler specific documentation. For example, this can be done with the -g switch in GCC. 

<b>SYM file via Breakpad</b>

Note that this approach should only be taken if it is impossible to provide the original binaries. Using .sym’s will result in less detailed information. 

To generate a .sym, you can use Breakpad’s dump_syms. Assuming the library in question is /lib/libfoo.so and its debugging symbol is /usr/debug/lib/libfoo.so, run: dump_syms /lib/libfoo.so /usr/debug/lib /tmp/libfoo.so.sym.
You can find additional info in [Breakpad’s documentation](https://www.chromium.org/developers/decoding-crash-dumps/).

</TabItem>
</Tabs>

## Generate symbol access tokens
Symbol access tokens are used to upload debug symbols. Go to Project Settings > Project > API Tokens to generate a token with the symbol:post capability. Additional details are available in [Creating API Tokens and Submission URLs](https://docs.saucelabs.com/error-reporting/project-setup/submission-url/#creating-api-tokens).

## Upload to Backtrace

### UI
Navigate to Project settings > Symbols > Overview to review your symbol uploads, 

> including upload history, search functionality, symbol access tokens and more. Backtrace recommends uploading archives (a .tar.gz or .zip file containing one or more .sym ,.pdb , ELF or dSYM files) of symbols for every build you expect to see crashes for. These symbol files can be uploaded manually or hooked up into your build system so they are automatically uploaded.

Reviewer: The commas in here are a little off:
`one or more .sym, .pdb, ELF or dSYM files`

> including upload history, search functionality, symbol access tokens and more. Backtrace recommends uploading archives (a .tar.gz or .zip file containing one or more .sym ,.pdb , ELF or dSYM files) of symbols for every build you expect to see crashes for. These symbol files can be uploaded manually or hooked up into your build system so they are automatically uploaded.

Reviewer: Maybe:
"These symbol files can be uploaded manually or by build automation scripts."

> including upload history, search functionality, symbol access tokens and more. Backtrace recommends uploading archives (a .tar.gz or .zip file containing one or more .sym ,.pdb , ELF or dSYM files) of symbols for every build you expect to see crashes for. These symbol files can be uploaded manually or hooked up into your build system so they are automatically uploaded.

Reviewer: > .sym ,.pdb , ELF or dSYM files
Should this also reference .so files?

You may also specify a tag to any of your symbol uploads in order to group symbols for ease of management. A tag is simply a group of symbols, like a folder on your filesystem. You may want to have different tags for different versions of your application or different platforms. If no tag is specified, symbols are placed into the anon tag.

### API

In order to build automation around symbol upload, such as integration into a build and release process, you'll want to interface directly with the HTTP API provided by Backtrace.

Below is an example of a curl command to submit a symbol archive.

```curl --data-binary @symbols.tar  -X POST  -H "Expect: gzip" "https://submit.backtrace.io/{your-subdomain}/{symbol-access-token}/symbols"```

Tags can be appended as a query parameter in the destination url (ie `"&tag=<tagValue>"`). 

> ### Supported formats

Reviewer: ```suggestion
### Supported archive formats
```

Any archive format supported by libarchive 3.2.3 is supported such as .tar, .tar.gz, .zip. Archives should be pre-compressed and be no larger than 10GB where possible. 

## Integrate custom symbol server with Backtrace

Instead of uploading symbols to Backtrace, users may choose to host a symbol server and upload symbols there. The symbol server can be integrated with Backtrace so that symbols are requested as reports requiring them are processed. 

symbold is the Backtrace service responsible for gathering symbols from connected symbol servers or stores. symbold comes pre configured by Backtrace to automatically download symbols from public third party symbol servers of commonly used libraries such as those from Microsoft, Electron, and others. We refer to these as default public symbol servers. You may also configure connections to your own private symbol servers to speed debugging time and minimize setup effort.

> See this [guide from Microsoft](https://docs.microsoft.com/en-us/windows/win32/debug/symbol-servers-and-symbol-stores) about setting up symbol servers and symbol stores.

Reviewer: ```suggestion
See this [Microsoft guide](https://docs.microsoft.com/en-us/windows/win32/debug/symbol-servers-and-symbol-stores) about setting up symbol servers and symbol stores.
```

- Have a web frontend to serve the files via HTTP(S)
   - Popular options include nginx, AWS, and Google Cloud buckets
- Network access from Backtrace to the symbol server
   - Firewall and network rules may need adjustments
- If compressing files, must use CAB mechanism
- [Optional] HTTP compression is also supported

### Feature details

Every time Backtrace receives a report, it is scanned for symbols that do not yet exist in that Backtrace project. A missing symbol event is generated for each of these symbols, and Backtrace then requests symbols from symbol servers accordingly.

Backtrace may request symbols from symbol servers for the following debug file, debug id, code file, and code id combinations, to support a variety of customer upload patterns:
- symbol_name = basename(debug_file); id = debug_id, if debug_file and debug _identifier are available

> - symbol_name = basename(code_file; id = code_id, if code_file and code_id are available and arch=x86_64

Reviewer: ```suggestion
- symbol_name = basename(code_file); id = code_id, if code_file and code_id are available and arch=x86_64
```

- base-url/symbol_name/id/symbol_name
- base-url/symbol_name/id/symbol_name_with_underscore_extension
- base-url/symbol_name/id/symbol_name_with_sym_extension

Where "symbol name with underscore extension" = replace last character of symbol name with underscore, '_' (this is typically used as the file extension for CAB compressed binaries on Windows) and where "symbol name with sym extension" = replace or append the extension ".sym" to the symbol name.

For example, if symbol_name = foo.pdb and id = 123, the following URLs are queried:
- base-url/foo.pdb/123/foo.pdb
- base-url/foo.pdb/123/foo.pd_
- base-url/foo.pdb/123/foo.sym

It is important that symbol servers only host valid symbol files. Symbolication may fail if multiple files exist in these paths, with one containing a valid symbol file and others are invalid. 

### Connecting Backtrace to symbol servers
You will need the following information to connect:
- URL - HTTPS URL to connect to the symbol server or S3 bucket. When using an AWS S3 bucket, use the HTTPS URL of the region the S3 bucket is hosted from.
- Name - A descriptive name for this connection.

> - Allowlist - Whether you want to enable Allowlist or not. If Allowlist is enabled, then the system will only try to download the symbols specified in the allowlist (no other symbols will be downloaded). 

Reviewer: ```suggestion
- Allowlist - If Allowlist is enabled, then the system will only try to download the symbols specified in the allowlist (no other symbols will be downloaded). 
```

- **Strongly recommended**: Allowlist - If the allowlist is enabled, then only symbol files in the allowlist will be downloaded. For the public symbol servers that are seeded by default, allowlists are enabled and commonly used symbols are listed for retrieval.
- Statistics and Usage - This tab show information about how much has been data has been downloaded, and how many successful or failed downloads there have been since the Symbol Server was added to the system
- Blocklist - Items in the blocklist will never be downloaded from the symbol server. Add items to the blocklist if they will not change stack trace information, will not add any additional debugging information, or might cause only networking problems (i.e files too big, they change too often, etc.).
- Skip list - Items are automatically added to the skip list if they cannot be downloaded from the symbol server in the specified number of retries (default is 3). Symbols in the skip list will not be requested. The number of retries can be configured. Skip list items over 12 hours old are flushed.
- Logs - Log information about successful and unsuccessful download and conversion attempts.

## Troubleshooting

> If you are blocked on uploading symbols from the command line, try uploading directly from your web browser. Instructions are available above.

Reviewer: Could this topic have a header, like there is an ### Invalid Token or ### Missing Symbols headers for the next two topics?

### Invalid Token

> If you receive an "invalid token" error in the response when uploading symbols via HTTP, check to make sure that you're using a symbol access token and not a submission token. To generate a symbol access token, see [Generating a Symbol Access Token](#generating-a-symbol-access-token).

Reviewer: ```suggestion
If you receive an "invalid token" error in the response when uploading symbols via HTTP, check to make sure that you're using a symbol access token and not a submission token. To generate a symbol access token, see [Generating Symbol Access Tokens](#generate-symbol-access-tokens).
```

### Missing symbols

See [Missing Symbols](https://docs.saucelabs.com/error-reporting/troubleshooting/#missing-symbols)

### Reading custom symbol server logs

> It’s important to note that some failures are expected, due to the wide net Backtrace casts (as explained in Symbol server | Feature details). To troubleshoot, search for the known good path on the symbol server (ie foo.pdb/123/foo.pdb). Successful symbol server requests log lines will include “Downloaded path”, and failed requests log lines will include “Download failed” with additional information about the error encountered.

Reviewer: ```suggestion
It’s important to note that some failures are expected, due to the wide net Backtrace casts (as explained in [Symbol server | Feature details](#feature-details)). To troubleshoot, search for the known good path on the symbol server (ie foo.pdb/123/foo.pdb). Successful symbol server requests log lines will include “Downloaded path”, and failed requests log lines will include “Download failed” with additional information about the error encountered.
```
